File: app/src/main/java/org/maroxa/diario/custom/ColorPickerView.java
package org.maroxa.diario.custom;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.SweepGradient;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class ColorPickerView extends View {
    private static final int[] COLORS = new int[]{
            Color.RED, Color.MAGENTA, Color.BLUE, Color.CYAN,
            Color.GREEN, Color.YELLOW, Color.RED
    };

    private Paint mPaint;
    private Paint mCenterPaint;
    private final int[] mColors;
    private boolean mTrackingCenter;
    private boolean mHighlightCenter;

    private int mCenterX;
    private int mCenterY;
    private int mCenterRadius = 32;
    private int mColorWheelRadius;
    private float brightness = 1.0f;

    private OnColorChangedListener mListener;

    public interface OnColorChangedListener {
        void colorChanged(int color);
    }

    public ColorPickerView(Context context) {
        this(context, null);
    }

    public ColorPickerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ColorPickerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mColors = COLORS;
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mCenterPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mCenterPaint.setColor(Color.RED);
        mCenterPaint.setStrokeWidth(5);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mPaint.setShader(new SweepGradient(mCenterX, mCenterY, mColors, null));

        canvas.drawCircle(mCenterX, mCenterY, mColorWheelRadius, mPaint);

        // Draw the center circle
        mCenterPaint.setStyle(Paint.Style.FILL);
        canvas.drawCircle(mCenterX, mCenterY, mCenterRadius, mCenterPaint);

        // Draw circle outline/highlight if center is being pressed
        if (mTrackingCenter) {
            int c = mCenterPaint.getColor();
            mCenterPaint.setStyle(Paint.Style.STROKE);
            if (mHighlightCenter) {
                mCenterPaint.setAlpha(0xFF);
            } else {
                mCenterPaint.setAlpha(0x80);
            }
            canvas.drawCircle(mCenterX, mCenterY, mCenterRadius + 5, mCenterPaint);
            mCenterPaint.setStyle(Paint.Style.FILL);
            mCenterPaint.setColor(c);
        }
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mCenterX = w / 2;
        mCenterY = h / 2;
        mColorWheelRadius = Math.min(mCenterX, mCenterY) - 10;
    }

    private int ave(int s, int d, float p) {
        return s + Math.round(p * (d - s));
    }

    private int interpColor(int[] colors, float unit) {
        if (unit <= 0) {
            return colors[0];
        }
        if (unit >= 1) {
            return colors[colors.length - 1];
        }

        float p = unit * (colors.length - 1);
        int i = (int) p;
        p -= i;

        int c0 = colors[i];
        int c1 = colors[i + 1];
        int a = ave(Color.alpha(c0), Color.alpha(c1), p);
        int r = ave(Color.red(c0), Color.red(c1), p);
        int g = ave(Color.green(c0), Color.green(c1), p);
        int b = ave(Color.blue(c0), Color.blue(c1), p);

        return Color.argb(a, r, g, b);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX() - mCenterX;
        float y = event.getY() - mCenterY;
        boolean inCenter = Math.sqrt(x * x + y * y) <= mCenterRadius;

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mTrackingCenter = inCenter;
                if (inCenter) {
                    mHighlightCenter = true;
                    invalidate();
                    break;
                }
            case MotionEvent.ACTION_MOVE:
                if (mTrackingCenter) {
                    if (mHighlightCenter != inCenter) {
                        mHighlightCenter = inCenter;
                        invalidate();
                    }
                } else {
                    float angle = (float) Math.atan2(y, x);
                    // Need to turn angle [-PI ... PI] into unit [0....1]
                    float unit = angle / (2 * (float) Math.PI);
                    if (unit < 0) {
                        unit += 1;
                    }
                    int color = interpColor(mColors, unit);
                    // Apply brightness
                    float[] hsv = new float[3];
                    Color.colorToHSV(color, hsv);
                    hsv[2] = brightness;
                    color = Color.HSVToColor(hsv);
                    
                    mCenterPaint.setColor(color);
                    if (mListener != null) {
                        mListener.colorChanged(color);
                    }
                    invalidate();
                }
                break;
            case MotionEvent.ACTION_UP:
                if (mTrackingCenter) {
                    if (inCenter) {
                        if (mListener != null) {
                            mListener.colorChanged(mCenterPaint.getColor());
                        }
                    }
                    mTrackingCenter = false;
                    invalidate();
                }
                break;
        }
        return true;
    }

    public void setOnColorChangedListener(OnColorChangedListener listener) {
        mListener = listener;
    }

    public void setColor(int color) {
        mCenterPaint.setColor(color);
        invalidate();
    }
    
    public int getColor() {
        return mCenterPaint.getColor();
    }
    
    public void setBrightness(float brightness) {
        this.brightness = brightness;
        int color = mCenterPaint.getColor();
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[2] = brightness;
        mCenterPaint.setColor(Color.HSVToColor(hsv));
        invalidate();
        
        if (mListener != null) {
            mListener.colorChanged(mCenterPaint.getColor());
        }
    }
} 

File: app/src/main/java/com/example/allinone/ui/drawing/ColorPickerView.kt
package com.example.allinone.ui.drawing

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.SweepGradient
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import kotlin.math.atan2
import kotlin.math.pow
import kotlin.math.sqrt

class ColorPickerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val centerPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val colorWheelRect = RectF()
    private var radius = 0f
    private var centerX = 0f
    private var centerY = 0f
    
    // Default center color is white
    private var centerColor = Color.WHITE
    
    private var onColorChangedListener: ((Int) -> Unit)? = null
    
    // Colors for the color wheel
    private val colors = intArrayOf(
        Color.RED, Color.MAGENTA, Color.BLUE, Color.CYAN,
        Color.GREEN, Color.YELLOW, Color.RED
    )

    init {
        centerPaint.color = centerColor
        centerPaint.style = Paint.Style.FILL
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        
        // Calculate the size of the color wheel
        val minDimension = minOf(w, h)
        radius = (minDimension / 2f) * 0.9f
        centerX = w / 2f
        centerY = h / 2f
        
        colorWheelRect.set(
            centerX - radius,
            centerY - radius,
            centerX + radius,
            centerY + radius
        )
        
        // Create the sweep gradient for the color wheel
        val shader = SweepGradient(centerX, centerY, colors, null)
        paint.shader = shader
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = radius / 4
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        
        // Draw the color wheel
        canvas.drawCircle(centerX, centerY, radius - paint.strokeWidth / 2, paint)
        
        // Draw the center circle with the selected color
        canvas.drawCircle(centerX, centerY, radius / 2, centerPaint)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val y = event.y
        
        when (event.action) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_MOVE -> {
                // Calculate the distance from center
                val dx = x - centerX
                val dy = y - centerY
                val distance = sqrt(dx.pow(2) + dy.pow(2))
                
                // Check if the touch is within the color wheel
                if (distance > radius / 2 && distance < radius) {
                    // Calculate the angle
                    val angle = (Math.toDegrees(atan2(dy, dx).toDouble()) + 360) % 360
                    
                    // Get the color at that angle
                    val hue = (angle / 360 * 360).toFloat()
                    val newColor = Color.HSVToColor(floatArrayOf(hue, 1f, 1f))
                    
                    // Update the center color
                    centerColor = newColor
                    centerPaint.color = centerColor
                    
                    // Notify the listener
                    onColorChangedListener?.invoke(centerColor)
                    
                    invalidate()
                    return true
                }
            }
        }
        
        return super.onTouchEvent(event)
    }
    
    fun setColor(color: Int) {
        centerColor = color
        centerPaint.color = color
        invalidate()
    }
    
    fun getColor(): Int {
        return centerColor
    }
    
    fun setOnColorChangedListener(listener: (Int) -> Unit) {
        onColorChangedListener = listener
    }
} 